File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.java
import java.util.*;

class Solution {
    public String solution(int n, int t, int m, String[] timetable) {
        ArrayList<Integer> timeList = new ArrayList<>();
        for (String time : timetable) {
            StringTokenizer st = new StringTokenizer(time, ":");
            int hour = Integer.parseInt(st.nextToken());
            int minute = Integer.parseInt(st.nextToken());
            timeList.add(hour * 60 + minute);
        }
        
        Collections.sort(timeList);
                
        int lastArvTime = 0;
        boolean isLastShuttleFull = false;
        for (int i = 0; i < n; i++) {
            int shtTime = 540 + i * t;
            int psgCnt = 0;
            
            while (timeList.size() > 0 && timeList.get(0) <= shtTime && psgCnt < m) {
                lastArvTime = timeList.remove(0);
                psgCnt++;
            }
            
            if (i == n - 1 && psgCnt == m) isLastShuttleFull = true;
        }
        
        int lastShtTime = (540 + (n - 1) * t >= 1440)? 1439 : 540 + (n - 1) * t;
        int conArvTime = (isLastShuttleFull)? lastArvTime - 1 : lastShtTime;
                
        return String.format("%02d:%02d", conArvTime / 60, conArvTime % 60);
    }
}

File: 프로그래머스/3/17678. ［1차］ 셔틀버스/［1차］ 셔틀버스.js
function solution(n, t, m, timetable) {
    timetable.sort();
    const minuteTimetable=timetable.map(a=>{
        return a.split(":")[0]*60+a.split(":")[1]*1;
    });

    const makeTime=(time)=>{
        let fhour=Math.floor(time/60);
        let hour=(fhour+"").length===2?fhour+"":("0"+fhour);
        let minute=(time%60+"").length===2?time%60+"":("0"+time%60);
        return hour+":"+minute;
    }

    const makeAnswer=()=>{
        let idx=0,myTime=0;
        for(let i=0;i<n;i++){
            let arriveTime=540+i*t;
            let count=0;
            myTime=arriveTime;
            if(i<n-1){
                for(let j=idx;j<minuteTimetable.length;j++){
                    if(minuteTimetable[j]<=arriveTime){
                        idx++; count++;
                    }   
                    if(count===m) break;
                }
            }
            else{
                for(let j=idx;j<minuteTimetable.length;j++){
                    if(minuteTimetable[j]<=arriveTime) count++;
                    if(count===m){
                        myTime=minuteTimetable[j]-1;
                        break;
                    }
                }
            }
        }

        return myTime;
    }

    //minuteTimetable.sort();
    return makeTime(makeAnswer());
}